amp/src/test/java/org/digijava/kernel/validators/activity/ImplementationLevelValidatorTest.java

```diff
@@ -0,0 +1,253 @@
+
+        AmpActivity activity = new AmpActivity();
+        activity.setCategories(ImmutableSet.of(regionalImplementationLevel, regionImplementationLocation));
+
+        Set<ConstraintViolation> violations = getConstraintViolations(activity);
+
+        assertThat(violations, emptyIterable());
+    }
+
+    @Test
+    public void testInvalidImplementationLocation() {
+        AmpCategoryValue centralImplLevel = categoryValues.getImplementationLevels().getCentral();
+        AmpCategoryValue zoneImplLocation = categoryValues.getImplementationLocations().getZone();
+
+        AmpActivity activity = new AmpActivity();
+        activity.setCategories(ImmutableSet.of(centralImplLevel, zoneImplLocation));
+
+        Set<ConstraintViolation> violations = getConstraintViolations(activity);
+
+        assertThat(violations, contains(implLocViolation(ActivityErrors.DOESNT_MATCH_IMPLEMENTATION_LEVEL)));
+    }
+
+    @Test
+    public void testMissingImplementationLocation() {
+        AmpCategoryValue centralImplLevel = categoryValues.getImplementationLevels().getCentral();
+
+        AmpActivity activity = new AmpActivity();
+        activity.setCategories(ImmutableSet.of(centralImplLevel));
+
+        Set<ConstraintViolation> violations = getConstraintViolations(activity);
+
+        assertThat(violations, emptyIterable());
+    }
+
+    @Test
+    public void testMissingImplementationLevel() {
+        AmpCategoryValue zoneImplLocation = categoryValues.getImplementationLocations().getZone();
+
+        AmpActivity activity = new AmpActivity();
+        activity.setCategories(ImmutableSet.of(zoneImplLocation));
+
+        Set<ConstraintViolation> violations = getConstraintViolations(activity);
+
+        assertThat(violations, contains(implLocViolation(ActivityErrors.IMPLEMENTATION_LEVEL_NOT_SPECIFIED)));
+    }
+
+    @Test
+    public void testMissingImplementationLevelAndLocation() {
+        AmpActivity activity = new AmpActivity();
+
+        Set<ConstraintViolation> violations = getConstraintViolations(activity);
+
+        assertThat(violations, emptyIterable());
+    }
+
+    //--- loc tests
+
+    @Test
+    public void testLocationAndImplLevelMismatch() {
+        AmpActivityLocation haitiUsage = new AmpActivityLocation();
+        haitiUsage.setLocation(locations.getAmpLocation("Haiti"));
+
+        AmpCategoryValue regionalImplementationLevel = categoryValues.getImplementationLevels().getRegional();
+        AmpCategoryValue regionImplementationLocation = categoryValues.getImplementationLocations().getRegion();
+
+        AmpActivity activity = new AmpActivity();
+        activity.setLocations(ImmutableSet.of(haitiUsage));
+        activity.setCategories(ImmutableSet.of(regionalImplementationLevel, regionImplementationLocation));
+
+        Set<ConstraintViolation> violations = getConstraintViolations(activity);
+
+        assertThat(violations, contains(locViolation(96L)));
```

Comment:
I understand this test for locations, though may be good to also check implLoc violation is reported together.

Reply:
If we're speaking of this particular test, then one violation is the correct answer.
If you mean to have a test which contains both kinds of violations at the same time then there is `#testInvalidLocationAndImplLoc`.

Reply:
You are right, I see it later in code